### PR TITLE
Mapper updates

### PR DESCRIPTION
**Pull request by rdaly525**
depends on #110 

Added a family_closure decorator along with tests for it.
Fixed bug regarding constructing the output type if there is a tuple
Fixed IR and min_pe to use family_closure
Added 'update_peak' function as an easier API to apply the ast_tools passes.

### CODE REVIEW

File: peak/mapper/mapper.py [hunk 1]
```diff
@@ -37,8 +40,15 @@ def __init__(self, peak_fc : tp.Callable):
         output_forms = []
         for input_form in input_forms:
             inputs = aadt_product_to_dict(input_form.value)
+
+            #Construct output_aadt value
             output = Peak_cls()(**inputs)
-            output = output_aadt_t(output)
+            if isinstance(output,tuple):
```
Comment (cdonovick) [inline peak/mapper/mapper.py]: whitespace

File: peak/peak.py [hunk 2]
```diff
@@ -66,7 +72,41 @@ def call_wrapper(*args,**kwargs):
...
+#2) Stores the family closure in Peak._fc_
+def family_closure(f):
+    num_inputs = f.__code__.co_argcount
+    if num_inputs != 1:
+        raise ValueError("Family Closure must take a single input 'family'")
+    @functools.lru_cache(None)
+    @functools.wraps(f)
+    def fc(family):
+        cls = f(family)
+        if not (isclass(cls) and issubclass(cls,Peak)):
```
Comment (cdonovick) [inline peak/peak.py]: whitespace

File: peak/peak.py [hunk 3]
```diff
@@ -66,7 +72,41 @@ def call_wrapper(*args,**kwargs):
...
+            raise ValueError("Family closure must return a peak class")
```
Comment (cdonovick) [inline peak/peak.py]: why?

File: peak/peak.py [hunk 4]
```diff
@@ -66,7 +72,41 @@ def call_wrapper(*args,**kwargs):
...
+def update_peak(peak_cls, family):
+    if family is SMTBit.get_family():
+        call = peak_cls.__call__
+        for dec in (
+            begin_rewrite(),
+            ssa(),
+            bool_to_bit(),
+            if_to_phi(family.Bit.ite),
+            end_rewrite()
+        ):
```
Comment (cdonovick) [inline peak/peak.py]: move to line of end_rewrite
python.org/dev/peps/pep-0008/#indentation

File: tests/test_interface.py [hunk 5]
```diff
@@ -23,3 +26,43 @@ def test_outputs():
         assert oname == expected_names[i]
         assert otype == expected_types[i]
 
+def test_family_closure():
+    #family_closure needs single argument
+    with pytest.raises(ValueError):
+        @family_closure
+        def fc(family,otherarg):
```
Comment (cdonovick) [inline tests/test_interface.py]: whitespace

File: tests/test_interface.py [hunk 6]
```diff
@@ -23,3 +26,43 @@ def test_outputs():
...
+            return 5
+        cls = fc(Bit.get_family())
+
+    #family_closure needs to return only a peak class
+    with pytest.raises(ValueError):
+        @family_closure
+        def fc(family):
+            class A(Peak): pass
+            return A, 5
+        cls,val = fc(Bit.get_family())
```
Comment (cdonovick) [inline tests/test_interface.py]: ws

File: tests/test_interface.py [hunk 7]
```diff
@@ -23,3 +26,43 @@ def test_outputs():
...
+    @family_closure
+    def PE_fc(family):
+        Word = family.BitVector[16]
+        class PE(Peak):
+            @name_outputs(out=Word)
+            def __call__(self,in0:Word,in1:Word):
+                return in0 + in1
+        return PE
+
+    assert hasattr(PE_fc,"_is_fc") and PE_fc._is_fc is True
```
Comment (cdonovick) [inline tests/test_interface.py]: ws

File: tests/test_ir.py [hunk 8]
```diff
@@ -35,6 +35,7 @@ def fun(family, a, b, c):
 
     assert "Simple" in ir.instructions
     Simple_fc = ir.instructions["Simple"]
+    assert hasattr(Simple_fc,"_is_fc") and Simple_fc._is_fc
```
Comment (cdonovick) [inline tests/test_ir.py]: ws

File: peak/mapper/mapper.py [hunk 9]
```diff
@@ -25,6 +26,8 @@
 
 class SMTMapper:
     def __init__(self, peak_fc : tp.Callable):
+        if not (hasattr(peak_fc,"_is_fc") and peak_fc._is_fc):
```
Comment (cdonovick) [inline peak/mapper/mapper.py]: whitespace

File: peak/ir.py [hunk 10]
```diff
@@ -49,7 +49,8 @@ def add_peak_instruction(self, name : str, input_interface : Product, output_int
             Peak=Peak,
             name_outputs=name_outputs,
             _fun_=fun,
-            _rebind_type=rebind_type
+            _rebind_type=rebind_type,
+            _family_closure=family_closure
         ))
         exec(class_src,exec_gs,exec_ls)
```
Comment (cdonovick) [inline peak/ir.py]: I know this isn't an addition but fix whitespace

File: tests/test_interface.py [hunk 11]
```diff
@@ -23,3 +26,43 @@ def test_outputs():
...
+            class A(Peak): pass
+            return A, 5
+        cls,val = fc(Bit.get_family())
+
+    @family_closure
+    def PE_fc(family):
+        Word = family.BitVector[16]
+        class PE(Peak):
+            @name_outputs(out=Word)
+            def __call__(self,in0:Word,in1:Word):
```
Comment (cdonovick) [inline tests/test_interface.py]: ws

File: peak/peak.py [hunk 12]
```diff
@@ -66,7 +72,41 @@ def call_wrapper(*args,**kwargs):
         return call_wrapper
     return decorator
 
+#This decorator does the following:
+#1) Caches the function call
+#2) Stores the family closure in Peak._fc_
+def family_closure(f):
+    num_inputs = f.__code__.co_argcount
+    if num_inputs != 1:
+        raise ValueError("Family Closure must take a single input 'family'")
```
Comment (cdonovick) [inline peak/peak.py]: Why? (I.e. why not wait till failure on line 85, if it is a failure)

File: peak/peak.py [hunk 13]
```diff
@@ -66,7 +72,41 @@ def call_wrapper(*args,**kwargs):
...
+    @functools.lru_cache(None)
+    @functools.wraps(f)
+    def fc(family):
+        cls = f(family)
+        if not (isclass(cls) and issubclass(cls,Peak)):
+            raise ValueError("Family closure must return a peak class")
+        cls._fc_ = fc
+        return cls
+    fc._is_fc = True
```
Comment (cdonovick) [inline peak/peak.py]: make `family_closure` a class, test with `isinstance`

File: peak/peak.py (hunk 3 above)
Comment (rdaly525) [reply peak/peak.py]: Not sure what you mean. I want to guarantee that the return value is a single peak class.

File: peak/peak.py (hunk 12 above)
Comment (rdaly525) [reply peak/peak.py]: There has been a lot of code written passing in family and generator parameters at the same time. I am preventing that in user code. For example not having the check would not prevent the following:

```
def gen_PE(family, width=16):
   ...
```

File: peak/peak.py (hunk 3 above)
Comment (cdonovick) [reply peak/peak.py]: Why do you want that?  Why is it necessary to inherit peak.  If everything works what does it matter if it inherits from peak?

File: peak/peak.py (hunk 12 above)
Comment (cdonovick) [reply peak/peak.py]: And?

File: peak/peak.py (hunk 3 above)
Comment (rdaly525) [reply peak/peak.py]: Inheriting from Peak implies I can get the input_t and output_t. Mapper does not work on non-Peak classes.

File: peak/peak.py (hunk 12 above)
Comment (rdaly525) [reply peak/peak.py]: Not sure what you are questioning. The point of family_closure is to force users to adopt a single style for writing peak classes that allows the auto-mapper to work. This style involves only passing in a single family as an argument.

File: peak/peak.py (hunk 3 above)
Comment (cdonovick) [reply peak/peak.py]: You still haven't told me any reason why we shouldn't use duck typing.  

If the mapper only works on classes whose call method have an input_t and output_t that is fine. Throw an error there.  Even mention that they should inherit from Peak.  However, if I manually build a class that looks like a Peak class that should be fine.  Unless there is some reason why it isn't.

File: peak/peak.py (hunk 12 above)
Comment (cdonovick) [reply peak/peak.py]: What breaks if you allow `def gen_PE(family, width=16):`?   The width might become an effectively irrelevant argument but what's the issue?

File: peak/peak.py [hunk 14]
```diff
@@ -33,11 +33,11 @@ def name_outputs(**outputs):
...
         @functools.wraps(call_fn)
-        def call_wrapper(*args,**kwargs):
-            results = call_fn(*args,**kwargs)
-            single_output = not isinstance(results,tuple)
+        def call_wrapper(*args, **kwargs):
+            results = call_fn(*args, **kwargs)
+            single_output = not isinstance(results, tuple)
             if single_output:
-                results = (results,)
+                results = (results, )
```
Comment (cdonovick) [inline peak/peak.py]: No white space after trailing comma
https://www.python.org/dev/peps/pep-0008/#pet-peeves

File: peak/peak.py (hunk 3 above)
Comment (rdaly525) [reply peak/peak.py]: Depends on your philosophy on building am embedded DSL. I am pro forcing users to adhere to a particular style that allows the DSL to work and to eagerly inform them with userful error messages when they inevitably screw up. This seems to be more important than allowing power users to unsafely ducktype. If you fundamentally disagree with this, then I'll change it. Maybe @phanrahan can comment.

Comment (cdonovick) [reply peak/peak.py]: I'd be fine with a warning.  But errors should be errors in my mind.

File: peak/peak.py (hunk 12 above)
Comment (cdonovick) [reply peak/peak.py]: I am fine with a warning here but seems unnecessary to make it an error.

Comment (cdonovick) [reply peak/peak.py]: Also you don't block `def gen_pe(family, *, width=16):`.  So I don't see why you should block `def gen_pe(family, width=16)`.

Comment (rdaly525) [reply peak/peak.py]: Changed to Warning

File: peak/peak.py (hunk 3 above)
Comment (rdaly525) [reply peak/peak.py]: Changed to Warning

File: tests/test_interface.py [hunk 15]
```diff
@@ -27,27 +27,6 @@ def test_outputs():
         assert otype == expected_types[i]
 
 def test_family_closure():
-    #family_closure needs single argument
-    with pytest.raises(SyntaxError):
```
Comment (cdonovick) [inline tests/test_interface.py]: If you want to keep these tests you can use `pytest.warns` 
https://docs.pytest.org/en/latest/warnings.html#warns

**Pull request merged**